pkg/syncer/cnsoperator/util/util.go

```diff
@@ -0,0 +1,103 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"context"
+	"fmt"
+
+	vmoperatortypes "github.com/vmware-tanzu/vm-operator-api/api/v1alpha1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/dynamic"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/vsphere-csi-driver/pkg/csi/service/logger"
+	k8s "sigs.k8s.io/vsphere-csi-driver/pkg/kubernetes"
+)
+
+var virtualNetworkGVR = schema.GroupVersionResource{
+	Group:    "vmware.com",
+	Version:  "v1alpha1",
+	Resource: "virtualnetworks",
+}
+
+const snatIPAnnotation = "ncp/snat_ip"
+
+// GetTKGVMIP finds the external facing IP address of a TKG VM object from a
+// given Supervisor Namespace based on the networking configuration (NSX-T or VDS)
+func GetTKGVMIP(ctx context.Context, vmOperatorClient client.Client, vmNamespace, vmName string, nsxConfiguration bool) (string, error) {
+	log := logger.GetLogger(ctx)
+	log.Infof("Determining external IP Address of VM: %s/%s", vmNamespace, vmName)
+	virtualMachineInstance := &vmoperatortypes.VirtualMachine{}
+	vmKey := types.NamespacedName{
+		Namespace: vmNamespace,
+		Name:      vmName,
+	}
+	err := vmOperatorClient.Get(ctx, vmKey, virtualMachineInstance)
+	if err != nil {
+		log.Errorf("failed to get virtualmachine %s/%s with error: %v", vmNamespace, vmName, err)
+		return "", err
+	}
+
+	var networkName string
+	for _, networkInterface := range virtualMachineInstance.Spec.NetworkInterfaces {
+		networkName = networkInterface.NetworkName
+	}
+	log.Debugf("VirtualMachine %s/%s is configured with network %s", vmNamespace, vmName, networkName)
+
+	var ip string
+	var exists bool
+	if nsxConfiguration {
+		virtualNetworkClient, err := getDynamicClientResourceInNamespace(ctx, vmNamespace, virtualNetworkGVR)
+		if err != nil {
+			return "", err
+		}
+		virtualNetworkInstance, err := virtualNetworkClient.Get(ctx, networkName, metav1.GetOptions{})
+		if err != nil {
+			return "", err
+		}
+		log.Debugf("Got VirtualNetwork instance %s/%s with annotations %v", vmNamespace, virtualNetworkInstance.GetName(), virtualNetworkInstance.GetAnnotations())
+		ip, exists = virtualNetworkInstance.GetAnnotations()[snatIPAnnotation]
+		if !exists {
+			return "", fmt.Errorf("failed to get SNAT IP annotation from VirtualMachine %s/%s", vmNamespace, vmName)
+		}
+	} else {
+		return "", fmt.Errorf("cannot determine external IP address of VDS based network configration")
```

Comment:
Do you want to put a TODO to add logic to find out the external facing IP of VMs in a VDS based Pacific?

Reply:
done